adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/X509AuthHandler.java

```diff
@@ -0,0 +1,127 @@
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.adapter.http;
+
+import java.net.HttpURLConnection;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Objects;
+
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.client.HonoClient;
+import org.eclipse.hono.service.auth.device.HonoAuthHandler;
+import org.eclipse.hono.service.auth.device.HonoClientBasedAuthProvider;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.eclipse.hono.util.RequestResponseApiConstants;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.impl.HttpStatusException;
+
+
+/**
+ * An authentication handler for client certificate based authentication.
+ *
+ */
+public class X509AuthHandler extends HonoAuthHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(X509AuthHandler.class);
+    private static final HttpStatusException UNAUTHORIZED = new HttpStatusException(HttpURLConnection.HTTP_UNAUTHORIZED);
+    private final HonoClient tenantServiceClient;
+
+    /**
+     * Creates a new handler for an authentication provider and a
+     * Tenant service client.
+     * 
+     * @param authProvider The authentication provider to use for verifying
+     *              the device identity.
+     * @param tenantServiceClient The client to use for determining the tenant
+     *              that the device belongs to.
+     */
+    public X509AuthHandler(final HonoClientBasedAuthProvider authProvider, final HonoClient tenantServiceClient) {
+        super(authProvider);
+        this.tenantServiceClient = Objects.requireNonNull(tenantServiceClient);
+    }
+
+    @Override
+    public final void parseCredentials(final RoutingContext context, final Handler<AsyncResult<JsonObject>> handler) {
+
+        Objects.requireNonNull(context);
+        Objects.requireNonNull(handler);
+
+        if (context.request().isSSL()) {
+            try {
+                final Certificate[] clientChain = context.request().sslSession().getPeerCertificates();
+                processCertificateChain(clientChain).setHandler(handler);
+            } catch (SSLPeerUnverifiedException e) {
+                // client certificate has not been validated
+                handler.handle(Future.failedFuture(UNAUTHORIZED));
+            }
+        } else {
+            handler.handle(Future.failedFuture(UNAUTHORIZED));
+        }
+    }
+
+    private Future<JsonObject> processCertificateChain(final Certificate[] clientChain) {
```

Comment:
That is true. However someone might want to implement checking other elements in the chain as well.

Reply:
*checking* as in: verify some properties and then fail the future if not as expected?

or are we talking about having access to the full chain in the getCredentials() method?

Reply:
My initial though that was latter. But it could be both.

It would be possible to check for other attributes as well, and fail if necessary. Like ruling out specific ciphers.

My understanding is that right now only the first element of the certificate chain is being verified. However it would be possible to walk up the chain as well.

For example could a device have a valid certificate, but based on their certificate chain, it might not be allowed to connect to this specific instance. Assume the protocol adapter is running in different regions. Devices get certificates signed with different intermediate certificates for different regions. Those device certificates are all valid. But are only allowed to connect to specific regions. In this case you would need access to the full chain.